src/libcrun/custom-handler.c

```diff
@@ -0,0 +1,595 @@
+    num_vcpus = CPU_COUNT (&set);
+
+  ctx_id = krun_create_ctx ();
+  if (UNLIKELY (ctx_id < 0))
+    error (EXIT_FAILURE, -ret, "could not create krun context");
+
+  ret = krun_set_vm_config (ctx_id, num_vcpus, ram_mib);
+  if (UNLIKELY (ret < 0))
+    error (EXIT_FAILURE, -ret, "could not set krun vm configuration");
+
+  ret = krun_set_root (ctx_id, "/");
+  if (UNLIKELY (ret < 0))
+    error (EXIT_FAILURE, -ret, "could not set krun root");
+
+  if (krun_set_workdir && def && def->process && def->process->cwd)
+    {
+      ret = krun_set_workdir (ctx_id, def->process->cwd);
+      if (UNLIKELY (ret < 0))
+        error (EXIT_FAILURE, -ret, "could not set krun working directory");
+    }
+
+  ret = krun_set_exec (ctx_id, pathname, &argv[1], NULL);
+  if (UNLIKELY (ret < 0))
+    error (EXIT_FAILURE, -ret, "could not set krun executable");
+
+  return krun_start_enter (ctx_id);
+}
+#endif
+
+#if HAVE_DLOPEN && HAVE_LIBKRUN
+/* libcrun_create_kvm_device: explicitly adds kvm device.  */
+static int
+libcrun_create_kvm_device (libcrun_context_t *context, libcrun_container_t *container, const char *rootfs, libcrun_error_t *err)
+{
+  int ret, rootfsfd;
+  size_t i;
+  struct device_s kvm_device = { "/dev/kvm", "c", 10, 232, 0666, 0, 0 };
+  cleanup_close int devfd = -1;
+  cleanup_close int rootfsfd_cleanup = -1;
+  runtime_spec_schema_config_schema *def = container->container_def;
+  const char *rootfs = get_private_data (container)->rootfs;
+  bool is_user_ns;
+
+  /* Do nothing if /dev/kvm is already present in spec */
+  for (i = 0; i < def->linux->devices_len; i++)
+    {
+      if (strcmp (def->linux->devices[i]->path, "/dev/kvm") == 0)
+        return 0;
+    }
+
+  if (rootfs == NULL)
+    rootfsfd = AT_FDCWD;
+  else
+    {
+      rootfsfd = rootfsfd_cleanup = open (rootfs, O_PATH);
+      if (UNLIKELY (rootfsfd < 0))
+        return crun_make_error (err, errno, "open `%s`", rootfs);
+    }
+  get_private_data (container)->rootfsfd = rootfsfd;
+  devfd = openat (rootfsfd, "dev", O_RDONLY | O_DIRECTORY);
+  if (UNLIKELY (devfd < 0))
+    return crun_make_error (err, errno, "open /dev directory in `%s`", rootfs);
+
+  is_user_ns = (get_private_data (container)->unshare_flags & CLONE_NEWUSER) ? true : false;
+
+  ret = create_dev (container, devfd, &kvm_device, is_user_ns, true, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  return 0;
+}
+#endif
+
+static int
+libcrun_configure_libkrun (struct custom_handler_s *out, libcrun_error_t *err)
+{
+#if HAVE_DLOPEN && HAVE_LIBKRUN
+  void *handle;
+#endif
+
+#if HAVE_DLOPEN && HAVE_LIBKRUN
+  handle = dlopen ("libkrun.so", RTLD_NOW);
+  if (handle == NULL)
+    return crun_make_error (err, 0, "could not load `libkrun.so`: %s", dlerror ());
+
+  out->exec_func = libkrun_do_exec;
+  out->exec_func_arg = handle;
+  out->post_configure_mounts = libcrun_create_kvm_device;
+
+  return 0;
+#else
+  (void) out;
+  return crun_make_error (err, ENOTSUP, "libkrun or dlopen not present");
+#endif
+}
+
```

Comment:
Same
```suggestion
if HAVE_DLOPEN && HAVE_WASMER
```

Reply:
I've moved the check inside the function so the function is always defined